packages/core/contracts/oracle/implementation/dvm2.0/Staker.sol

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../Voting.sol";
+import "../../../common/implementation/Testable.sol";
+
+contract Staker is Voting {
+    using ResultComputation for ResultComputation.Data;
+    // Each User Stake is tracked with the information below.
+
+    // APY emission trackers
+    uint256 emissionRate = 634200000000000000; // ~20% emission per year on 100mm tokens.
+    uint256 cumulativeStaked;
+    uint256 rewardPerTokenStored;
+    uint256 lastUpdateTime;
+
+    // Slashing trackers
+    uint256 lastRequestIndexConsidered;
+
+    struct SlashingTracker {
+        uint256 wrongVoteSlashPerToken;
+        uint256 noVoteSlashPerToken;
+        uint256 totalSlashed;
+    }
+
+    SlashingTracker[] requestSlashTrackers;
+
+    uint256 actionLockTime = 30 days;
+
+    struct VoterStake {
+        uint256 cumulativeStaked;
+        // Pro-rata trackers.
+        uint256 rewardsPaidPerToken;
+        uint256 rewardsOutstanding;
+        uint256 unstakeTime;
+        uint256 requestUnstake;
+        // Slashing trackers.
+        uint256 lastRequestIndexConsidered;
+        // Negative represents loss due to slashing. Positive represents gains due to voting right and capturing others slashed amounts.
+        int256 unrealizedSlash;
+    }
+
+    mapping(address => VoterStake) stakingBalances;
+
+    constructor(
+        uint256 _phaseLength,
+        FixedPoint.Unsigned memory _gatPercentage,
+        FixedPoint.Unsigned memory _inflationRate,
+        uint256 _rewardsExpirationTimeout,
+        address _votingToken,
+        address _finder,
+        address _timerAddress
+    )
+        Voting(
+            _phaseLength,
+            _gatPercentage,
+            _inflationRate,
+            _rewardsExpirationTimeout,
+            _votingToken,
+            _finder,
+            _timerAddress
+        )
+    {}
+
+    // Pulls tokens from users wallet and stakes them.
+    function stake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+
+        stakingBalances[msg.sender].cumulativeStaked += amount;
+        cumulativeStaked += amount;
+
+        if (stakingBalances[msg.sender].lastRequestIndexConsidered == 0)
+            stakingBalances[msg.sender].lastRequestIndexConsidered = priceRequestIds.length;
+
+        votingToken.transferFrom(msg.sender, address(this), amount);
+    }
+
+    function requestUnstake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+        // Staker signals that they want to unstake. After signalling, their total voting balance is decreased by the
+        // signaled amount. This amount is not vulnerable to being slashed but also does not accumulate rewards.
+
+        require(stakingBalances[msg.sender].cumulativeStaked >= amount, "Bad request amount");
+
+        stakingBalances[msg.sender].requestUnstake = amount;
```

Comment:
This method allows a user to restake as long as they haven't withdrawn, right? i.e. I call with amount = 10. Later, I call this method with amount = 0. My 10 is now staked again.

Without snapshotting the requestUnstake value at some point in the voting process, this would (I think) open a loophole for someone to unstake all their tokens during the vote and have the option to restake before rewards are assessed once they have more information about the result of the vote.